doc/src/build/programming-with-objects/ch4-object-wrapping.md

```diff
@@ -0,0 +1,195 @@
+
+The full source code can be found in [TrustedSwap.move](../../../../sui_programmability/examples/objects_tutorial/sources/TrustedSwap.move).
+
+A more complex example of using direct wrapping can be found in [Escrow.move](../../../../sui_programmability/examples/defi/sources/Escrow.move).
+
+### Wrapping through `Option`
+When Sui object type `Bar` is directly wrapped into `Foo`, there is not much flexiblity: a `Foo` object must has a `Bar` object in it, and in order to take out the `Bar` object one must destroy the `Foo` object. However, there are cases where we want more flexibility: the wrapping type may or may not always has the wrapped object in it, and the wrapped object may be replaced with a different object at some point.
+
+Let's demonstrate this use case by designing a simple game character: A warrior with a sword and shield. A warrior may or may not have a sword and shield, and they should be able to replace them anytime. To design this, we define a `SimpleWarrior` type as following:
+```rust
+struct SimpleWarrior has key {
+    id: VersionedID,
+    sword: Option<Sword>,
+    shield: Option<Shield>,
+}
+```
+Each `SimpleWarrior` type has an optional `sword` and `shield` wrapped in it, defined as:
+```rust
+struct Sword has key, store {
+    id: VersionedID,
+    strength: u8,
+}
+
+struct Shield has key, store {
+    id: VersionedID,
+    armor: u8,
+}
+```
+When we are creating a new warrior, we can set the `sword` and `shield` to none to indicate that there are no equipment yet:
+```rust
+public(script) fun create_warrior(ctx: &mut TxContext) {
+    let warrior = SimpleWarrior {
+        id: TxContext::new_id(ctx),
+        sword: Option::none(),
+        shield: Option::none(),
+    };
+    Transfer::transfer(warrior, TxContext::sender(ctx))
+}
+```
+With this, we can then define functions to equip new swords or new shields:
+```rust
+public(script) fun equip_sword(warrior: &mut SimpleWarrior, sword: Sword, ctx: &mut TxContext) {
+    if (Option::is_some(&warrior.sword)) {
+        let old_sword = Option::extract(&mut warrior.sword);
+        Transfer::transfer(old_sword, TxContext::sender(ctx));
+    };
+    Option::fill(&mut warrior.sword, sword);
+}
+```
+In the above function, we are passing a `warrior` as mutable reference of `SimpleWarrior`, and a `sword` passed by value because we need to wrap it into the `warrior`.
+
+It is important to note that because `Sword` is a Sui object type without `drop` ability, if the warrior already has a sword equipped, that sword cannot just be dropped. If we make a call to `Option::fill` without first checking and taking out the existig sword, a runtime error may happen. Hence in `equip_sword`, we first check if there is already a sword equipped, and if so, we take it out and send it back to the sender. This matches what you would typically expect when you equip a new sword: if a sword is already equipped, you get it back.
```

Comment:
```suggestion
It is important to note that because `Sword` is a Sui object type without `drop` ability, if the warrior already has a sword equipped, that sword cannot just be dropped. If we make a call to `Option::fill` without first checking and taking out the existig sword, a runtime error may happen. Hence in `equip_sword`, we first check if there is already a sword equipped, and if so, we take it out and send it back to the sender. This matches what you would expect when you equip a new sword--you get the old sword back, if there is one.
```